plugins/kubernetes/app/models/watchers/base_watcher.rb

```diff
@@ -8,18 +8,24 @@ class BaseWatcher
     def initialize(watch_stream, log: true)
       @watch_stream, @log, = watch_stream, log
       @start_time = Time.now
+      @enabled = false
     end
 
     def start_watching(&block)
+      @original_thread_id = Thread.current.object_id
+      @enabled = true
       @watch_stream.each do |notice|
         handle_notice(notice, &block)
+        break if stop_watching?
       end
+      cleanup
     end
 
     def stop_watching
-      if @watch_stream
-        @watch_stream.finish
-        @watch_stream = nil
+      if @enabled
+        @enabled = false
+        # close the socket and stop receiving updates if called from a different thread than the original
+        cleanup if Thread.current.object_id != @original_thread_id
```

Comment:
In the current state of the plugin there are no other threads involved. However, if we were to have a supervisor thread, or any other thread trying to stop the watcher when the watcher was stuck waiting for events, it would have no effect. The 2nd commit is addressing this (currently hypothetical) scenario.